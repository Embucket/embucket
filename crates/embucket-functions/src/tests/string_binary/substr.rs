use crate::test_query;

test_query!(
    substr_basic_positive,
    "SELECT substr('mystring', 3, 2) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_basic_positive_no_length,
    "SELECT substr('mystring', 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_negative_index_multiple,
    "SELECT substr('mystring', -3, 2) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_negative_index_no_length,
    "SELECT substr('mystring', -3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_zero_position,
    "SELECT substr('mystring', 0, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_out_of_bounds_positive,
    "SELECT substr('mystring', 20, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_out_of_bounds_negative,
    "SELECT substr('mystring', -20, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_empty_string,
    "SELECT substr('', 1, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_empty_string_negative,
    "SELECT substr('', -1, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_null_string,
    "SELECT substr(NULL, 1, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_null_start,
    "SELECT substr('mystring', NULL, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_null_length,
    "SELECT substr('mystring', 1, NULL) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_zero_length,
    "SELECT substr('mystring', 3, 0) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_unicode_chars,
    "SELECT substr('h√©lloüåèworld', 3, 4) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_unicode_negative,
    "SELECT substr('h√©lloüåèworld', -3, 2) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_long_string,
    "SELECT substr('this is a very long string for testing purposes', 10, 8) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_long_string_negative,
    "SELECT substr('this is a very long string for testing purposes', -8, 5) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_edge_case_one_char,
    "SELECT substr('a', 1, 1) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_edge_case_one_char_negative,
    "SELECT substr('a', -1, 1) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_multiple_rows,
    "SELECT substr(col, 2, 3) as result FROM (VALUES ('hello'), ('world'), ('test')) AS t(col)",
    snapshot_path = "substr"
);

test_query!(
    substr_multiple_rows_negative,
    "SELECT substr(col, -2, 2) as result FROM (VALUES ('hello'), ('world'), ('test')) AS t(col)",
    snapshot_path = "substr"
);

test_query!(
    substr_negative_length_error,
    "SELECT substr('mystring', 1, -1) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_large_length,
    "SELECT substr('mystring', 3, 100) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_large_negative_start,
    "SELECT substr('mystring', -100, 5) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_integer_positive,
    "SELECT substr(12345, 2, 3) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_integer_negative,
    "SELECT substr(-567, -2, 2) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_decimal_positive,
    "SELECT substr(123.456, 3, 4) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_decimal_negative_index,
    "SELECT substr(1.23, -2, 2) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_binary_to_binary_basic,
    "SELECT substr(to_binary('hello world'), 2, 5) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_binary_to_binary_negative,
    "SELECT substr(to_binary('hello world'), -5, 4) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_binary_to_binary_emoji,
    "SELECT substr(to_binary('testüöÄdata'), 5, 4) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_binary_to_binary_emoji_negative,
    "SELECT substr(to_binary('testüöÄdata'), -8, 4) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_binary_to_binary_mixed,
    "SELECT substr(to_binary('caf√©‚òïtest'), 4, 5) as result",
    snapshot_path = "substr"
);

test_query!(
    substr_binary_to_binary_empty,
    "SELECT substr(to_binary(''), 1, 3) as result",
    snapshot_path = "substr"
);
