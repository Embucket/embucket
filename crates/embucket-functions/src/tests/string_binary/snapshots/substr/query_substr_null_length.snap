---
source: crates/embucket-functions/src/tests/string_binary/substr.rs
description: "\"SELECT substr('mystring', 1, NULL) as result\""
---
Ok(
    [
        "+--------+",
        "| result |",
        "+--------+",
        "|        |",
        "+--------+",
    ],
)
